File: src/main/java/cookiedragon/luchadora/event/api/EventDispatcher.java
package cookiedragon.luchadora.event.api;

import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

/**
 * <p>An annotated event bus implementation</p>
 *
 * <p>It is designed with the basis that classes are registered once, indexing their methods for potential event subscribers.</p>
 *
 * <p>This means the class can be frequently subscribed and unsubscribed with minimal performance impact</p>
 *
 * <p>A limitation of this event bus implementation is that subscribed methods cannot be inherited from super classes.
 * I might add this at a later date</p>
 *
 * @author cookiedragon234 07/Dec/2019
 */
public class EventDispatcher
{
	private static final Map<Class<?>, Set<SubscribingMethod>> subscriptions = new ConcurrentHashMap<>();
	
	/**
	 * <p>This will index the given class, searching it for methods that subscribe to events. This will **NOT**
	 * subscribe the class to receive events. In order to that you need to, well, subscribe it! Just use the
	 * subscribe function and pass the same class.</p>
	 *
	 * <p><b>This will only index non-static methods</b></p>
	 *
	 * @param subscriber An initialised class containing methods to be subscribed
	 */
	public static void register(Object subscriber)
	{
		register(subscriber, subscriber.getClass());
	}
	
	/**
	 * <p>This will index the given class, searching it for methods that subscribe to events. This will **NOT**
	 * subscribe the class to receive events. In order to that you need to, well, subscribe it! Just use the
	 * subscribe function and pass the same class.</p>
	 *
	 * <p><b>This will only index static methods</b></p>
	 *
	 * @param clazz The class to index
	 */
	public static void register(Class clazz)
	{
		// We provide null as the subscriber instance because it is indexing static methods and therefore needs no instance to initialise
		register(null, clazz);
	}
	
	private static void register(Object subscriber, Class clazz)
	{
		for (Method declaredMethod : clazz.getDeclaredMethods())
		{
			// If we are registering a static class then only allow static methods to be indexed!
			if (subscriber == null && !Modifier.isStatic(declaredMethod.getModifiers()))
				continue;
			
			// On the other hand, if we are indexing an initialised class then skip static methods!
			if (subscriber != null && Modifier.isStatic(declaredMethod.getModifiers()))
				continue;
			
			if (!declaredMethod.isAnnotationPresent(Subscriber.class))
				continue;
			
			if (declaredMethod.getParameterCount() != 1)
				throw new RuntimeException("Subscribing event must have only 1 parameter!");
			
			Class<?> eventType = declaredMethod.getParameterTypes()[0];
			
			if (!AbstractEvent.class.isAssignableFrom(eventType))
				throw new RuntimeException(String.format("Registered method '%s' parameter needs to extends AbstractEvent", declaredMethod.toString()));
			
			if (!subscriptions.containsKey(eventType))
				subscriptions.put(eventType, new HashSet<>());
			
			subscriptions.get(eventType).add(new SubscribingMethod(clazz, subscriber, declaredMethod));
		}
	}
	
	/**
	 * Subscribes the given class, meaning that it's methods will be eligible to recieve events.
	 * You **MUST** register the class before subscribing it.
	 *
	 * @param subscriber The class to subscribe
	 */
	public static void subscribe(Class subscriber)
	{
		for (Set<SubscribingMethod> subscribingMethods : subscriptions.values())
		{
			for (SubscribingMethod subscribingMethod : subscribingMethods)
			{
				if (subscribingMethod.clazz == subscriber)
				{
					subscribingMethod.active = true;
				}
			}
		}
	}
	
	/**
	 * Subscribes the given class, meaning that it's methods will be eligible to recieve events.
	 * You **MUST** register the class before subscribing it.
	 *
	 * @param subscriber The class to subscribe
	 */
	public static void subscribe(Object subscriber)
	{
		for (Set<SubscribingMethod> subscribingMethods : subscriptions.values())
		{
			for (SubscribingMethod subscribingMethod : subscribingMethods)
			{
				if (subscribingMethod.instance == subscriber)
				{
					subscribingMethod.active = true;
				}
			}
		}
	}
	
	/**
	 * Unsubscribe the class, therefore making it ineligible to receive events.
	 *
	 * @param subscriber The class to subscribe
	 */
	public static void unsubscribe(Class subscriber)
	{
		for (Set<SubscribingMethod> subscribingMethods : subscriptions.values())
		{
			for (SubscribingMethod subscribingMethod : subscribingMethods)
			{
				if (subscribingMethod.clazz == subscriber)
				{
					subscribingMethod.active = false;
				}
			}
		}
	}
	
	/**
	 * Unsubscribe the class, therefore making it ineligible to receive events.
	 *
	 * @param subscriber The class to subscribe
	 */
	public static void unsubscribe(Object subscriber)
	{
		for (Set<SubscribingMethod> subscribingMethods : subscriptions.values())
		{
			for (SubscribingMethod subscribingMethod : subscribingMethods)
			{
				if (subscribingMethod.instance == subscriber)
				{
					subscribingMethod.active = false;
				}
			}
		}
	}
	
	/**
	 * Dispatches the given event to all the methods that subscribe to it
	 *
	 * @param event The event to dispatch
	 * @return The event that was dispatched
	 */
	public static AbstractEvent dispatch(AbstractEvent event)
	{
		for (Map.Entry<Class<?>, Set<SubscribingMethod>> entry : subscriptions.entrySet())
		{
			if (entry.getKey().isAssignableFrom(event.getClass()))
			{
				Iterator<SubscribingMethod> iterator = entry.getValue().iterator();
				while (iterator.hasNext() && !event.isCancelled())
				{
					SubscribingMethod subscribingMethod = iterator.next();
					if (subscribingMethod.active)
					{
						try
						{
							subscribingMethod.invoke(event);
						}
						catch(Exception e)
						{
							throw new RuntimeException("Error while invoking event '" + event.getClass().getName() + "'", e);
						}
					}
				}
			}
		}
		
		return event;
	}
}

File: src/main/kotlin/cookiedragon/eventsystem/EventDispatcher.kt
package cookiedragon.eventsystem

/**
 * @author cookiedragon234 15/Feb/2020
 */
interface EventDispatcher {
	/**
	 * Dispatches the given event to any registered and subscribed listeners targeting
	 * this event's class or subclasses
	 *
	 * @param event The event to dispatch
	 * @return The same event♦
	 */
	fun <T: Any> dispatch(event: T): T
	
	
	/**
	 * Registers *static* methods of the given class
	 *
	 * During registration methods of the given class are indexed, but not subscribed
	 * Subscribers will not receive events until a registered class is subscribed
	 *
	 * @param subscriber An uninitialised class
	 */
	fun register(subscriber: Class<*>)
	
	/**
	 * Registers *non static* methods of the given object
	 *
	 * During registration methods of the given object are indexed, but not subscribed
	 * Subscribers will not receive events until a registered object is subscribed
	 *
	 * @param subscriber An initialised object
	 */
	fun register(subscriber: Any)
	
	
	/**
	 * Subscribes *static* methods of the given class
	 *
	 * After subscription any registered methods within the given class will receive events
	 *
	 * @param subscriber An uninitialised and registered class
	 */
	fun subscribe(subscriber: Class<*>)
	
	/**
	 * Subscribes *non static* methods of the given object
	 *
	 * After subscription any registered methods within the given object will receive events
	 *
	 * @param subscriber An initialised and registered object
	 */
	fun subscribe(subscriber: Any)
	
	
	/**
	 * Unsubscribes *static* methods of the given class
	 *
	 * After unsubscription any registered methods within the given class will *no longer* receive events
	 *
	 * @param subscriber An uninitialised and registered class
	 */
	fun unsubscribe(subscriber: Class<*>)
	
	/**
	 * Unsubscribes *non static* methods of the given object
	 *
	 * After unsubscription any registered methods within the given object will *no longer* receive events
	 *
	 * @param subscriber An initialised and registered object
	 */
	fun unsubscribe(subscriber: Any)
	
	
	companion object: EventDispatcher {
		override fun <T : Any> dispatch(event: T): T =      EventDispatcherImpl.dispatch(event)
		
		override fun register(subscriber: Class<*>) =       EventDispatcherImpl.register(subscriber)
		override fun register(subscriber: Any) =            EventDispatcherImpl.register(subscriber)
		
		override fun subscribe(subscriber: Class<*>) =      EventDispatcherImpl.subscribe(subscriber)
		override fun subscribe(subscriber: Any) =           EventDispatcherImpl.subscribe(subscriber)
		
		override fun unsubscribe(subscriber: Class<*>) =    EventDispatcherImpl.unsubscribe(subscriber)
		override fun unsubscribe(subscriber: Any) =         EventDispatcherImpl.unsubscribe(subscriber)
	}
}
